### PR TITLE
Add the timestamp of last received zigbee message visible in HASS

### PR DESCRIPTION
**Pull request by ugrug**
- can be enabled in settings / advanced / add_timestamp: true
- on enabled add last_message json attribute to mqtt payload

### CODE REVIEW

File: lib/extension/homeassistant.js [hunk 1]
```diff
@@ -465,6 +465,16 @@ class HomeAssistant {
             // Set unique_id
             payload.unique_id = `${ieeeAddr}_${config.object_id}_${settings.get().mqtt.base_topic}`;
 
+            // Add last_message to attributes
+            if (settings.get().advanced.add_timestamp) {
+                if (!payload.hasOwnProperty('json_attributes')) {
+                    payload.json_attributes = [];
+                }
+                if (!payload.json_attributes.includes('last_message')) {
+                    payload.json_attributes.push('last_message');
```
[the remark this answers is not in the document]
Comment (ugrug) [reply lib/extension/homeassistant.js]: After checking the source, seems currently only (https://www.home-assistant.io/components/sensor.mqtt/) supports `json_attributes`.

File: lib/extension/deviceReceive.js [hunk 2]
```diff
@@ -99,6 +99,11 @@ class DeviceReceive {
                     payload.linkquality = message.linkquality;
                 }
 
+                // Add last message received
+                if (settings.get().advanced.add_timestamp) {
+                    payload.last_message = Date.now();
```
[the remark this answers is not in the document]
Comment (ugrug) [reply lib/extension/deviceReceive.js]: You are right, renamed.

**Comment by ugrug** [comment]
After removed the option, check failing now as messages different now from previous values.

File: lib/extension/deviceReceive.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (ugrug) [reply lib/extension/deviceReceive.js]: What do you think about:
- saving last_seen in device (`device.last_seen`) ?
It should be great if `Controller.prototype.checkOnline` can be used to check router capable device availability, this case this timestamp can be modified from there on endpoint just receiving messages. Started a pull request for insert callback function, hopefully will be accepted.
- use `payload.last_seen = new Date(device.last_seen).toISOString()` in payload for readable format ?

**Comment by ugrug** [comment]
Done :)

**Comment by Koenkk** [comment]
Thanks :bee:

**Comment by Koenkk** [comment]
FYI last_seen now requires configuration, see: https://github.com/Koenkk/zigbee2mqtt/blob/dev/docs/configuration/configuration.md

**Pull request merged**